File: 155-min-stack/min-stack.js

var MinStack = function() {
    this.stack=[];
    this.min=[]
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    
    this.stack.push(val);
    if(this.min.length===0){
        this.min.push(val);
    }
    else if(this.min[this.min.length-1]>=val){
        this.min.push(val)
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    let num=this.stack.pop();
    if(num===this.min[this.min.length-1]){
        this.min.pop()
    }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.stack[this.stack.length-1]
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.min[this.min.length-1];
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: 155-min-stack/min-stack.py
class MinStack(object):

    def __init__(self):
        self.stack=[]
        self.minstack=[]

    def push(self, val):
        self.stack.append(val)
        if self.minstack==[]:
            self.minstack.append(val)
        else:
            self.minstack.append(min(val,self.minstack[-1]))
        

    def pop(self):
       self.stack.pop()
       self.minstack.pop()
        

    def top(self):
       return self.stack[-1]
        

    def getMin(self):
        return self.minstack[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()